### PR TITLE
Improve model checkpoint saving logic

### PR DESCRIPTION
**Pull request by lucas-ventura**
Improve model checkpoint saving logic to always save model checkpoint when validation is not run

# What does this PR do?
This PR enhances the model checkpoint saving logic to always save the model checkpoint when validation is not run. Previously, the model was only saved if validation was performed and the validation loss improved. With this change, we ensure that a checkpoint is saved after every epoch if validation is turned off (``--run_validation False``).

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: src/llama_recipes/utils/train_utils.py [hunk 1]
```diff
@@ -220,72 +220,78 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
 
         # Update the learning rate as needed
         lr_scheduler.step()
+        should_save_model = False
```
Comment (mreso) [inline src/llama_recipes/utils/train_utils.py:223]:
```suggestion
        should_save_model = train_config.save_model
```

File: src/llama_recipes/utils/train_utils.py [hunk 2]
```diff
@@ -220,72 +220,78 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
         if train_config.run_validation:
             eval_ppl, eval_epoch_loss, temp_val_loss, temp_step_perplexity = evaluation(model, train_config, eval_dataloader, local_rank, tokenizer, wandb_run)
             if train_config.save_metrics:
                 val_step_loss.extend(temp_val_loss)
                 val_step_perplexity.extend(temp_step_perplexity)
-
+            should_save_model = eval_epoch_loss < best_val_loss
+        else:
+            should_save_model = True
```
Comment (mreso) [inline src/llama_recipes/utils/train_utils.py:231]:
```suggestion
            should_save_model = train_config.save_model and eval_epoch_loss < best_val_loss
```

**Review by mreso — COMMENTED** [review]
Hi, thanks for the contribution! Made a small change so we still follow the train_config.save_model parameter.  Are you able to you add the logs of some tests runs that show the logic paths work?

**Comment by lucas-ventura** [comment]
Hi!

I tried with ``--run_validation False`` and ``--run_validation True`` it worked for both cases.

File: src/llama_recipes/utils/train_utils.py [hunk 3]
```diff
@@ -220,72 +220,76 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
         if train_config.run_validation:
             eval_ppl, eval_epoch_loss, temp_val_loss, temp_step_perplexity = evaluation(model, train_config, eval_dataloader, local_rank, tokenizer, wandb_run)
             if train_config.save_metrics:
                 val_step_loss.extend(temp_val_loss)
                 val_step_perplexity.extend(temp_step_perplexity)
-
+            should_save_model = train_config.save_model and eval_epoch_loss < best_val_loss
+        
+        if should_save_model:
             checkpoint_start_time = time.perf_counter()
```
Comment (mreso) [inline src/llama_recipes/utils/train_utils.py:232]:
```suggestion
        checkpoint_start_time = time.perf_counter()
        if should_save_model:
```

File: src/llama_recipes/utils/train_utils.py [hunk 4]
```diff
@@ -220,72 +220,76 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
+                    else:
+                        print(" Saving the FSDP model checkpoints and optimizer using SHARDED_STATE_DICT")
+                        print("=====================================================")
+                        save_model_and_optimizer_sharded(model, rank, train_config)
+
+                    
+            if train_config.enable_fsdp:
+                dist.barrier()
             checkpoint_end_time = time.perf_counter() - checkpoint_start_time
             checkpoint_times.append(checkpoint_end_time)
```
Comment (mreso) [inline src/llama_recipes/utils/train_utils.py:281]:
```suggestion
        checkpoint_end_time = time.perf_counter() - checkpoint_start_time
        checkpoint_times.append(checkpoint_end_time)
```

File: src/llama_recipes/utils/train_utils.py [hunk 5]
```diff
@@ -220,72 +220,76 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
-                if train_config.enable_fsdp:
-                    if rank==0:
+
+            if train_config.run_validation:
+                if eval_epoch_loss < best_val_loss:
+                    best_val_loss = eval_epoch_loss
+                    if train_config.enable_fsdp:
+                        if rank==0:
+                            print(f"best eval loss on epoch {epoch+1} is {best_val_loss}")
+                    else:
```
Comment (mreso) [inline src/llama_recipes/utils/train_utils.py:289]:
```suggestion

        if train_config.run_validation:
            if eval_epoch_loss < best_val_loss:
                best_val_loss = eval_epoch_loss
                if train_config.enable_fsdp:
                    if rank==0:
                        print(f"best eval loss on epoch {epoch+1} is {best_val_loss}")
                else:
```

File: src/llama_recipes/utils/train_utils.py [hunk 6]
```diff
@@ -220,72 +220,76 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
                         print(f"best eval loss on epoch {epoch+1} is {best_val_loss}")
-                else:
-                    print(f"best eval loss on epoch {epoch+1} is {best_val_loss}")
-            val_loss.append(float(best_val_loss))
-            val_prep.append(float(eval_ppl))
+                val_loss.append(float(best_val_loss))
+                val_prep.append(float(eval_ppl))
```
Comment (mreso) [inline src/llama_recipes/utils/train_utils.py:292]:
```suggestion
            val_loss.append(float(best_val_loss))
            val_prep.append(float(eval_ppl))
```

**Review by mreso — COMMENTED** [review]
Did some more tests and run_validation = True and save_model = False does result in division through zero.

**Comment by mreso** [comment]
Command
```
torchrun --nnodes 1 --nproc_per_node 2  recipes/quickstart/finetuning/finetuning.py --enable_fsdp --model_name  meta-llama/Llama-3.1-8B-Instruct --use_peft --peft_method lora --output_dir output --samsum_dataset.trust_remote_code=True --train_config.max_train_step 1 --run_validation True --save_model True
```
Output
```
Training Epoch: 1:   0%|                                                                                                                                                                                                                                                                                                                           | 0/79 [00:00<?, ?it/s]
NCCL version 2.21.5+cuda12.4
Training Epoch: 1/3, step 0/79 completed (loss: 1.4149080514907837):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:09<12:56,  9.96s/it]
max training steps reached, stopping training, total train steps finished:  1
Training Epoch: 1/3, step 0/79 completed (loss: 1.4149080514907837):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:10<13:08, 10.10s/it]
Training Epoch: 1/3, step 0/79 completed (loss: 1.4178253412246704):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:11<15:26, 11.88s/it]
Max CUDA memory allocated was 37 GB
Max CUDA memory reserved was 44 GB
Peak active CUDA memory was 37 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 7 GB
evaluating Epoch: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17/17 [00:08<00:00,  2.04it/s]
evaluating Epoch: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17/17 [00:08<00:00,  2.03it/s]
 eval_ppl=tensor(4.1797, device='cuda:0') eval_epoch_loss=tensor(1.4303, device='cuda:0')
we are about to save the PEFT modules
PEFT modules are saved in output directory
Starting epoch 1/3
train_config.max_train_step: 1
best eval loss on epoch 1 is 1.4302507638931274
Epoch 1: train_perplexity=1.0181, train_epoch_loss=0.0179, epoch time 11.075395782478154s
Starting epoch 1/3
train_config.max_train_step: 1
Key: avg_train_prep, Value: 1.0180903673171997
Key: avg_train_loss, Value: 0.01792869344353676
Key: avg_eval_prep, Value: 4.179747581481934
Key: avg_eval_loss, Value: 1.4302507638931274
Key: avg_epoch_time, Value: 11.075395782478154
Key: avg_checkpoint_time, Value: 28.332108918577433
```
Command
```
torchrun --nnodes 1 --nproc_per_node 2  recipes/quickstart/finetuning/finetuning.py --enable_fsdp --model_name  meta-llama/Llama-3.1-8B-Instruct --use_peft --peft_method lora --output_dir output --samsum_dataset.trust_remote_code=True --train_config.max_train_step 1 --run_validation False --save_model True
```
Output
```
Training Epoch: 1:   0%|                                                                                                                                                                                                                                                                                                                           | 0/79 [00:00<?, ?it/s]
NCCL version 2.21.5+cuda12.4
Training Epoch: 1/3, step 0/79 completed (loss: 1.4149080514907837):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:09<12:58,  9.98s/it]
max training steps reached, stopping training, total train steps finished:  1
Training Epoch: 1/3, step 0/79 completed (loss: 1.4149080514907837):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:10<13:05, 10.07s/it]
Training Epoch: 1/3, step 0/79 completed (loss: 1.4178253412246704):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:13<17:03, 13.12s/it]
Max CUDA memory allocated was 37 GB
Max CUDA memory reserved was 44 GB
Peak active CUDA memory was 37 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 7 GB
we are about to save the PEFT modules
PEFT modules are saved in output directory
Starting epoch 1/3
train_config.max_train_step: 1
Epoch 1: train_perplexity=1.0181, train_epoch_loss=0.0179, epoch time 11.032661393284798s
Starting epoch 1/3
train_config.max_train_step: 1
Key: avg_train_prep, Value: 1.0180903673171997
Key: avg_train_loss, Value: 0.01792869344353676
Key: avg_epoch_time, Value: 11.032661393284798
Key: avg_checkpoint_time, Value: 25.04800534900278
```
Command
```
torchrun --nnodes 1 --nproc_per_node 2  recipes/quickstart/finetuning/finetuning.py --enable_fsdp --model_name  meta-llama/Llama-3.1-8B-Instruct --use_peft --peft_method lora --output_dir output --samsum_dataset.trust_remote_code=True --train_config.max_train_step 1 --run_validation True --save_model False
```
Output
```
Training Epoch: 1:   0%|                                                                                                                                                                                                                                                                                                                           | 0/79 [00:00<?, ?it/s]
NCCL version 2.21.5+cuda12.4
Training Epoch: 1/3, step 0/79 completed (loss: 1.4149080514907837):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:11<14:22, 11.06s/it]
Training Epoch: 1:   1%|███▉                                                                                                                                                                                                                                                                                                               | 1/79 [00:11<14:38, 11.26s/it]
Training Epoch: 1/3, step 0/79 completed (loss: 1.4149080514907837):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:11<14:29, 11.15s/it]
Training Epoch: 1/3, step 0/79 completed (loss: 1.4178253412246704):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:11<14:45, 11.35s/it]
Max CUDA memory allocated was 37 GB
Max CUDA memory reserved was 44 GB
Peak active CUDA memory was 37 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 7 GB
evaluating Epoch: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17/17 [00:08<00:00,  2.04it/s]
evaluating Epoch: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17/17 [00:08<00:00,  2.02it/s]
Starting epoch 1/3
train_config.max_train_step: 1
 eval_ppl=tensor(4.1801, device='cuda:0') eval_epoch_loss=tensor(1.4303, device='cuda:0')
best eval loss on epoch 1 is 1.4303467273712158
Epoch 1: train_perplexity=1.0181, train_epoch_loss=0.0179, epoch time 11.879832149483263s
Starting epoch 1/3
train_config.max_train_step: 1
Key: avg_train_prep, Value: 1.0180903673171997
Key: avg_train_loss, Value: 0.01792869344353676
Key: avg_eval_prep, Value: 4.180148601531982
Key: avg_eval_loss, Value: 1.4303467273712158
Key: avg_epoch_time, Value: 11.879832149483263
Key: avg_checkpoint_time, Value: 5.047768354415894e-07```

```
Command
```
torchrun --nnodes 1 --nproc_per_node 2  recipes/quickstart/finetuning/finetuning.py --enable_fsdp --model_name  meta-llama/Llama-3.1-8B-Instruct --use_peft --peft_method lora --output_dir output --samsum_dataset.trust_remote_code=True --train_config.max_train_step 1 --run_validation False --save_model False
```
Output
```
Training Epoch: 1:   0%|                                                                                                                                                                                                                                                                                                                           | 0/79 [00:00<?, ?it/s]
NCCL version 2.21.5+cuda12.4
Training Epoch: 1/3, step 0/79 completed (loss: 1.4149080514907837):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:08<11:41,  8.99s/it]
max training steps reached, stopping training, total train steps finished:  1
Training Epoch: 1/3, step 0/79 completed (loss: 1.4178253412246704):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:10<13:50, 10.65s/it]
Training Epoch: 1/3, step 0/79 completed (loss: 1.4149080514907837):   1%|███▎                                                                                                                                                                                                                                                             | 1/79 [00:09<11:48,  9.09s/it]
Starting epoch 1/3
train_config.max_train_step: 1
Max CUDA memory allocated was 37 GB
Max CUDA memory reserved was 44 GB
Peak active CUDA memory was 37 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 7 GB
Epoch 1: train_perplexity=1.0181, train_epoch_loss=0.0179, epoch time 9.935478443279862s
Starting epoch 1/3
train_config.max_train_step: 1
Key: avg_train_prep, Value: 1.0180903673171997
Key: avg_train_loss, Value: 0.01792869344353676
Key: avg_epoch_time, Value: 9.935478443279862
Key: avg_checkpoint_time, Value: 4.880130290985107e-07
```
Command
```
python recipes/quickstart/finetuning/finetuning.py  --model_name  meta-llama/Llama-3.1-8B-Instruct --use_peft --peft_method lora --output_dir output --samsum_dataset.trust_remote_code=True --train_config.max_train_step 1 --quantization 8bit --run_validation True --save_model True
```
Output
```
Training Epoch: 1/3, step 0/159 completed (loss: 1.5974043607711792):   1%|█▌                                                                                                                                                                                                                                                             | 1/159 [00:14<37:04, 14.08s/it]
max training steps reached, stopping training, total train steps finished:  1
Training Epoch: 1/3, step 0/159 completed (loss: 1.5974043607711792):   1%|█▌                                                                                                                                                                                                                                                             | 1/159 [00:14<37:50, 14.37s/it]
Max CUDA memory allocated was 42 GB
Max CUDA memory reserved was 47 GB
Peak active CUDA memory was 42 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 16 GB
evaluating Epoch: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 34/34 [00:27<00:00,  1.24it/s]
 eval_ppl=tensor(4.2325, device='cuda:0') eval_epoch_loss=tensor(1.4428, device='cuda:0')
we are about to save the PEFT modules
PEFT modules are saved in output directory
best eval loss on epoch 1 is 1.442801594734192
Epoch 1: train_perplexity=1.0101, train_epoch_loss=0.0100, epoch time 15.41115855704993s
Starting epoch 1/3
train_config.max_train_step: 1
Key: avg_train_prep, Value: 1.0100972652435303
Key: avg_train_loss, Value: 0.010046567767858505
Key: avg_eval_prep, Value: 4.232537269592285
Key: avg_eval_loss, Value: 1.442801594734192
Key: avg_epoch_time, Value: 15.41115855704993
Key: avg_checkpoint_time, Value: 0.2108828453347087
```

Command
```
python recipes/quickstart/finetuning/finetuning.py  --model_name  meta-llama/Llama-3.1-8B-Instruct --use_peft --peft_method lora --output_dir output --samsum_dataset.trust_remote_code=True --train_config.max_train_step 1 --quantization 8bit --run_validation False --save_model True
```
Output
```
Training Epoch: 1:   0%|                                                                                                                                                                                                                                                                                                                          | 0/159 [00:00<?, ?it/s]
/home/mreso/.conda/envs/llama/lib/python3.10/site-packages/bitsandbytes/autograd/_functions.py:316: UserWarning: MatMul8bitLt: inputs will be cast from torch.bfloat16 to float16 during quantization
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
Training Epoch: 1/3, step 0/159 completed (loss: 1.5974043607711792):   1%|█▌                                                                                                                                                                                                                                                             | 1/159 [00:14<38:39, 14.68s/it]
max training steps reached, stopping training, total train steps finished:  1
Training Epoch: 1/3, step 0/159 completed (loss: 1.5974043607711792):   1%|█▌                                                                                                                                                                                                                                                             | 1/159 [00:14<39:09, 14.87s/it]
Max CUDA memory allocated was 42 GB
Max CUDA memory reserved was 47 GB
Peak active CUDA memory was 42 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 16 GB
we are about to save the PEFT modules
PEFT modules are saved in output directory
Epoch 1: train_perplexity=1.0101, train_epoch_loss=0.0100, epoch time 15.737907895818353s
Starting epoch 1/3
train_config.max_train_step: 1
Key: avg_train_prep, Value: 1.0100972652435303
Key: avg_train_loss, Value: 0.010046567767858505
Key: avg_epoch_time, Value: 15.737907895818353
Key: avg_checkpoint_time, Value: 0.21278224047273397
```

Command
```
python recipes/quickstart/finetuning/finetuning.py  --model_name  meta-llama/Llama-3.1-8B-Instruct --use_peft --peft_method lora --output_dir output --samsum_dataset.trust_remote_code=True --train_config.max_train_step 1 --quantization 8bit --run_validation True --save_model False
```
Output
```
Training Epoch: 1/3, step 0/159 completed (loss: 1.5974043607711792):   1%|█▌                                                                                                                                                                                                                                                             | 1/159 [00:13<36:13, 13.76s/it]
max training steps reached, stopping training, total train steps finished:  1
Training Epoch: 1/3, step 0/159 completed (loss: 1.5974043607711792):   1%|█▌                                                                                                                                                                                                                                                             | 1/159 [00:14<36:54, 14.01s/it]
Max CUDA memory allocated was 42 GB
Max CUDA memory reserved was 47 GB
Peak active CUDA memory was 42 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 16 GB
evaluating Epoch: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 34/34 [00:28<00:00,  1.21it/s]
 eval_ppl=tensor(4.2355, device='cuda:0') eval_epoch_loss=tensor(1.4435, device='cuda:0')
best eval loss on epoch 1 is 1.4434908628463745
Epoch 1: train_perplexity=1.0101, train_epoch_loss=0.0100, epoch time 16.009447498247027s
Starting epoch 1/3
train_config.max_train_step: 1
Key: avg_train_prep, Value: 1.0100972652435303
Key: avg_train_loss, Value: 0.010046567767858505
Key: avg_eval_prep, Value: 4.235455513000488
Key: avg_eval_loss, Value: 1.4434908628463745
Key: avg_epoch_time, Value: 16.009447498247027
Key: avg_checkpoint_time, Value: 5.662441253662109e-07
```

Command
```
python recipes/quickstart/finetuning/finetuning.py  --model_name  meta-llama/Llama-3.1-8B-Instruct --use_peft --peft_method lora --output_dir output --samsum_dataset.trust_remote_code=True --train_config.max_train_step 1 --quantization 8bit --run_validation False --save_model False
```
Output
```
Training Epoch: 1:   0%|                                                                                                                                                                                                                                                                                                                          | 0/159 [00:00<?, ?it/s]
/home/mreso/.conda/envs/llama/lib/python3.10/site-packages/bitsandbytes/autograd/_functions.py:316: UserWarning: MatMul8bitLt: inputs will be cast from torch.bfloat16 to float16 during quantization
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
Training Epoch: 1/3, step 0/159 completed (loss: 1.5974043607711792):   1%|█▌                                                                                                                                                                                                                                                             | 1/159 [00:14<37:06, 14.09s/it]
max training steps reached, stopping training, total train steps finished:  1
Training Epoch: 1/3, step 0/159 completed (loss: 1.5974043607711792):   1%|█▌                                                                                                                                                                                                                                                             | 1/159 [00:14<37:48, 14.36s/it]
Max CUDA memory allocated was 42 GB
Max CUDA memory reserved was 47 GB
Peak active CUDA memory was 42 GB
CUDA Malloc retries : 0
CPU Total Peak Memory consumed during the train (max): 16 GB
Epoch 1: train_perplexity=1.0101, train_epoch_loss=0.0100, epoch time 15.381424403749406s
Starting epoch 1/3
train_config.max_train_step: 1
Key: avg_train_prep, Value: 1.0100972652435303
Key: avg_train_loss, Value: 0.010046567767858505
Key: avg_epoch_time, Value: 15.381424403749406
Key: avg_checkpoint_time, Value: 3.734603524208069e-07

```

**Review by mreso — APPROVED** [review]
Now it look good to me! Thanks a lot for contributing!
